File: examples/node/test/worker.test.js
const test = require('node:test');
const assert = require('node:assert/strict');

function freshModules() {
    delete require.cache[require.resolve('../lib/repos')];
    delete require.cache[require.resolve('../lib/parse_report')];
    delete require.cache[require.resolve('../lib/queue')];
    delete require.cache[require.resolve('../worker')];
    return {
        repos: require('../lib/repos'),
        worker: require('../worker'),
    };
}

const primary = {
    id: 'rpt-primary-1',
    time: '2026-04-09T10:47:58.612Z',
    device: { id: 'dev-primary', displayId: 'HGD-1', model: { family: 'locoTrack', name: 'HGD4', product: 'LocoTrack', version: 4, revision: 1 }, labels: [] },
    owner: { id: 'own-1', name: 'Acme' },
    location: { type: 'wifi', global: { lat: 1, lon: 2 } },
    sensors: { battery: { level: 90 }, temperature: 20 },
    secondaryObservations: [
        { id: 'dev-tag-A', displayId: 'TAG-A', model: { family: 'locoCard', name: 'E1BL', product: 'LocoCard', version: 1, revision: 0 }, reportType: 'presence', reportId: 'rpt-tag-A', rssi: -50, isPaired: true },
        { id: 'dev-tag-B', displayId: 'TAG-B', model: { family: 'locoCard', name: 'E1BL', product: 'LocoCard', version: 1, revision: 0 }, reportType: 'event',    reportId: 'rpt-tag-B', rssi: -70, isPaired: true },
    ],
};

const secondary = {
    id: 'rpt-tag-A',
    time: '2026-04-09T10:48:51.000Z',
    device: { id: 'dev-tag-A', displayId: 'TAG-A', model: { family: 'locoCard', name: 'E1BL', product: 'LocoCard', version: 1, revision: 0 }, labels: [] },
    owner: { id: 'own-1', name: 'Acme' },
    reportedBy: {
        device: { id: 'dev-primary', displayId: 'HGD-1', model: { family: 'locoTrack', name: 'HGD4', product: 'LocoTrack', version: 4, revision: 1 } },
        report: { id: 'rpt-primary-1', time: '2026-04-09T10:47:58.612Z' },
        rssi: -50,
    },
    sensors: { battery: { level: 80 } },
    timeSeries: { temperature: [{ time: '2026-04-09T10:36:00.000Z', value: 22.6 }] },
};

test('worker stores the primary report and presence-touches each secondary', async () => {
    const { repos, worker } = freshModules();
    repos._resetAll();

    await worker.handleDeviceReport(primary);

    const stored = repos.deviceReports._get('rpt-primary-1');
    assert.equal(stored.role, 'primary');
    assert.equal(stored.secondaries.length, 2);

    const primaryDev = repos.devices._get('dev-primary');
    assert.equal(primaryDev.lastReport.id, 'rpt-primary-1');
    assert.equal(primaryDev.lastReport.role, 'primary');
    assert.equal(primaryDev.lastSensors.battery.level, 90);

    // Both secondaries should have a presence touch — but NOT a lastSensors
    // entry, because the primary's view is presence-only.
    const tagA = repos.devices._get('dev-tag-A');
    const tagB = repos.devices._get('dev-tag-B');
    assert.equal(tagA.lastSeenBy.primaryDeviceId, 'dev-primary');
    assert.equal(tagA.lastSeenBy.rssi, -50);
    assert.equal(tagB.lastSeenBy.primaryDeviceId, 'dev-primary');
    assert.equal(tagB.lastSeenBy.rssi, -70);
    assert.ok(!('lastSensors' in tagA), 'primary presence touch must not write sensor data');
});

test('worker stores the secondary report with authoritative tag sensors', async () => {
    const { repos, worker } = freshModules();
    repos._resetAll();

    // Order doesn't matter; deliveries arrive independently per the guide.
    await worker.handleDeviceReport(secondary);
    await worker.handleDeviceReport(primary);

    const tagA = repos.devices._get('dev-tag-A');
    // Tag's own report wrote sensors before the primary's presence touch
    // landed. The primary's touch should not clobber lastSensors.
    assert.equal(tagA.lastSensors.battery.level, 80);
    assert.equal(tagA.lastSeenBy.primaryDeviceId, 'dev-primary');
    assert.equal(tagA.lastReport.id, 'rpt-tag-A');
    assert.equal(tagA.lastReport.role, 'secondary');
});

test('worker falls back to the relay primary when the tag has no shipment binding', async () => {
    const { repos, worker } = freshModules();
    repos._resetAll();
    repos.shipments._seed('ship-7', { boundDeviceIds: ['dev-primary'] });

    await worker.handleDeviceReport(secondary);

    const ship = repos.shipments._get('ship-7');
    assert.equal(ship.events.length, 1);
    assert.equal(ship.events[0]._id, 'rpt-tag-A');
});

test('worker appends to a shipment matched by device name (the conventional shipment carrier)', async () => {
    const { repos, worker } = freshModules();
    repos._resetAll();
    repos.shipments._seed('ship-42', { boundDeviceNames: ['SHIP-42'] });

    await worker.handleDeviceReport({
        ...primary,
        id: 'rpt-with-name',
        device: { ...primary.device, name: 'SHIP-42' },
    });

    const ship = repos.shipments._get('ship-42');
    assert.equal(ship.events.length, 1);
});

test('worker is null-safe when the report has nothing optional', async () => {
    const { repos, worker } = freshModules();
    repos._resetAll();
    await worker.handleDeviceReport({
        id: 'rpt-empty',
        time: '2026-04-09T10:00:00.000Z',
        device: { id: 'dev-empty', displayId: 'X', model: { family: 'locoTrack', name: 'HGD4', product: 'LocoTrack', version: 4, revision: 1 }, labels: [] },
        owner: { id: 'own-1', name: 'Acme' },
    });
    const stored = repos.deviceReports._get('rpt-empty');
    assert.equal(stored.role, 'standalone');
    assert.equal(repos.devices._get('dev-empty')._id, 'dev-empty');
});

File: examples/ruby/test/worker_test.rb
require_relative 'test_helper'
require_relative '../worker'

class WorkerTest < Minitest::Test
  PRIMARY = {
    'id' => 'rpt-primary-1',
    'time' => '2026-04-09T10:47:58.612Z',
    'device' => {
      'id' => 'dev-primary', 'displayId' => 'HGD-1',
      'model' => { 'family' => 'locoTrack', 'name' => 'HGD4', 'product' => 'LocoTrack', 'version' => 4, 'revision' => 1 },
      'labels' => []
    },
    'owner' => { 'id' => 'own-1', 'name' => 'Acme' },
    'sensors' => { 'battery' => { 'level' => 90 }, 'temperature' => 20 },
    'secondaryObservations' => [
      { 'id' => 'dev-tag-A', 'displayId' => 'TAG-A',
        'model' => { 'family' => 'locoCard', 'name' => 'E1BL', 'product' => 'LocoCard', 'version' => 1, 'revision' => 0 },
        'reportType' => 'presence', 'reportId' => 'rpt-tag-A', 'rssi' => -50, 'isPaired' => true },
      { 'id' => 'dev-tag-B', 'displayId' => 'TAG-B',
        'model' => { 'family' => 'locoCard', 'name' => 'E1BL', 'product' => 'LocoCard', 'version' => 1, 'revision' => 0 },
        'reportType' => 'event', 'reportId' => 'rpt-tag-B', 'rssi' => -70, 'isPaired' => true }
    ]
  }.freeze

  SECONDARY = {
    'id' => 'rpt-tag-A',
    'time' => '2026-04-09T10:48:51.000Z',
    'device' => {
      'id' => 'dev-tag-A', 'displayId' => 'TAG-A',
      'model' => { 'family' => 'locoCard', 'name' => 'E1BL', 'product' => 'LocoCard', 'version' => 1, 'revision' => 0 },
      'labels' => []
    },
    'owner' => { 'id' => 'own-1', 'name' => 'Acme' },
    'reportedBy' => {
      'device' => { 'id' => 'dev-primary', 'displayId' => 'HGD-1',
                    'model' => { 'family' => 'locoTrack', 'name' => 'HGD4', 'product' => 'LocoTrack', 'version' => 4, 'revision' => 1 } },
      'report' => { 'id' => 'rpt-primary-1', 'time' => '2026-04-09T10:47:58.612Z' },
      'rssi' => -50
    },
    'sensors' => { 'battery' => { 'level' => 80 } }
  }.freeze

  def setup
    DocStore.reset_all
  end

  def test_primary_stores_report_and_touches_each_secondary
    Worker.handle_device_report(PRIMARY)

    stored = DeviceReports.get('rpt-primary-1')
    assert_equal 'primary', stored['role']
    assert_equal 2, stored['secondaries'].size

    primary_dev = Devices.get('dev-primary')
    assert_equal 'rpt-primary-1', primary_dev['lastReport']['id']
    assert_equal 90, primary_dev['lastSensors']['battery']['level']

    tag_a = Devices.get('dev-tag-A')
    tag_b = Devices.get('dev-tag-B')
    assert_equal 'dev-primary', tag_a['lastSeenBy']['primaryDeviceId']
    assert_equal(-50, tag_a['lastSeenBy']['rssi'])
    assert_equal(-70, tag_b['lastSeenBy']['rssi'])
    refute tag_a.key?('lastSensors'), 'primary presence touch must not write sensor data'
  end

  def test_secondary_writes_authoritative_sensors_and_primary_does_not_clobber
    Worker.handle_device_report(SECONDARY)
    Worker.handle_device_report(PRIMARY)

    tag_a = Devices.get('dev-tag-A')
    assert_equal 80, tag_a['lastSensors']['battery']['level']
    assert_equal 'dev-primary', tag_a['lastSeenBy']['primaryDeviceId']
    assert_equal 'rpt-tag-A', tag_a['lastReport']['id']
    assert_equal 'secondary', tag_a['lastReport']['role']
  end

  def test_worker_falls_back_to_relay_primary_when_tag_has_no_shipment_binding
    Shipments.seed('ship-7', 'boundDeviceIds' => ['dev-primary'])

    Worker.handle_device_report(SECONDARY)

    ship = Shipments.get('ship-7')
    assert_equal 1, ship['events'].size
    assert_equal 'rpt-tag-A', ship['events'][0]['_id']
  end

  def test_worker_appends_to_shipment_matched_by_device_name
    Shipments.seed('ship-42', 'boundDeviceNames' => ['SHIP-42'])

    payload = PRIMARY.merge(
      'id' => 'rpt-with-name',
      'device' => PRIMARY['device'].merge('name' => 'SHIP-42')
    )
    Worker.handle_device_report(payload)

    ship = Shipments.get('ship-42')
    assert_equal 1, ship['events'].size
  end

  def test_worker_is_null_safe_for_a_minimal_report
    Worker.handle_device_report(
      'id' => 'rpt-empty',
      'time' => '2026-04-09T10:00:00.000Z',
      'device' => {
        'id' => 'dev-empty', 'displayId' => 'X',
        'model' => { 'family' => 'locoTrack', 'name' => 'HGD4', 'product' => 'LocoTrack', 'version' => 4, 'revision' => 1 },
        'labels' => []
      },
      'owner' => { 'id' => 'own-1', 'name' => 'Acme' }
    )

    stored = DeviceReports.get('rpt-empty')
    assert_equal 'standalone', stored['role']
    assert_equal 'dev-empty', Devices.get('dev-empty')['_id']
  end
end
